PR for the deploy.sh file

namitshah opened the pull request:
This will add the deploy.sh file to the primary testbed-develop branch.

namitshah committed:
Added deploy.sh

namitshah committed:
Update deploy.sh

Fixed a minor bug.

babcs2035 commented on WAFL-Testbed/ctrl/deploy.sh:
```diff
@@ -0,0 +1,54 @@
+    echo "Directory names do not match!"
+    exit 1
+fi
+if [ ! -d "$target_confirm" ] 
```
If `target_confirm` does not exist, you may want to create the directory.

babcs2035 commented on WAFL-Testbed/ctrl/deploy.sh:
```diff
@@ -0,0 +1,54 @@
+mkdir -p "$target_confirm/wafl/dataset"
+mkdir -p "$target_confirm/wafl/config"
+mkdir -p "$target_confirm/wafl/src"
+final_location="/home/denjo/testbed/"
```
By the usage of `final_location` and itself, `target_location` may be useless?
You may want to use `target_location` for the later processes.

babcs2035 commented on WAFL-Testbed/ctrl/deploy.sh:
```diff
@@ -0,0 +1,54 @@
+echo "Directories will have the following path: $final_location$target_name"
+
+# The Devices of the Network
+devices=("192.168.11.100" "192.168.11.101" "192.168.11.102" 
+        "192.168.11.103" "192.168.11.104" "192.168.11.105" 
+        "192.168.11.106" "192.168.11.107" "192.168.11.108" 
+        "192.168.11.109")
```
`devices` should be defined by environment variables, which are defined by wafl_execution_base_config.

babcs2035 commented on WAFL-Testbed/ctrl/deploy.sh:
```diff
@@ -0,0 +1,54 @@
+        "192.168.11.103" "192.168.11.104" "192.168.11.105" 
+        "192.168.11.106" "192.168.11.107" "192.168.11.108" 
+        "192.168.11.109")
+user="denjo"
```
Should we define the user name of each server as an environment variable?

babcs2035 reviewed (commented):
I left some comments on the script, can you check them?

nat0391 commented on WAFL-Testbed/ctrl/deploy.sh:
```diff
@@ -0,0 +1,54 @@
+    scp -r -q "$target_confirm/wafl/src" "$user@$device:$final_location$target_name"
+    ssh "$user@$device" "mkdir -p $final_location$target_name/results"
+    echo "Successfully deployed project to the device"
+done
```
This script copies the same directory to all devices, but according to the document, it seems that each device is expected to be able to have different sources, datasets, and configs.
Could you check 4.1.1~4.1.6 of the document?

namitshah committed:
Updated the shell script and the base config file

namitshah commented:
Hi, I have fixed the issues. Please review the updated files.
Some of the Changes:
-> Will use the current directory as the project directory.
Please run the script from the project directory (as ./ctrl/deploy.sh). Removed the inputs and confirmation section.
-> Have ONLY changed the list structure of the base execution config file, so that deserialization with bash is easier, have not modified any of the contents.
-> Every device will have its own directory for device-specific data, and the data from the common sub-directory will be overwritten by device-specific sub-directory data.
-> the base execution config file will also be sent to all the devices (as per the documentation).

Please let me know if any other changes are required.
Thank you.

Ta2305 commented:
To use environment variables exported in `wafl_base_execution_config`, the following command may be needed.
`source "$TARGET_PATH/ctrl/wafl_base_execution_config"`

namitshah commented:
> To use environment variables exported in `wafl_base_execution_config`, the following command may be needed. `source "$TARGET_PATH/ctrl/wafl_base_execution_config"`

I thought that  the source command is to be run separately when initializing before running the shell script. I'll add the statement to the script.

namitshah committed:
Update deploy.sh

Minor update (regarding environment variable import).

nat0391 commented:
I’ve reviewed the code and found no issues.

namitshah commented:
Since the script seems to cover all the requirements defined in the documentation, I've added a few basic functionalities to it: SSH timeout, exception handling, unsuccessful deployment detection (ctrl/unsuccessful_deployment_list.txt will be generated in case an error is encountered), and support for partially successful deployment. 
In case deployment to a few devices fails because of some network / configuration issue, the script will keep running for the other devices, and will generate a list of failed deployments, so that troubleshooting can be done and the script can be run again only for those specific devices.

Please let me know if the PR can now be merged.
Thank you.

namitshah committed:
Update deploy.sh

Added support for partially successful deployment

nat0391 commented on WAFL-Testbed/ctrl/deploy.sh:
```diff
@@ -0,0 +1,72 @@
+    # The Base Configuration shell script is also sent to the execution servers
+    scp -r -q "$TARGET_PATH/ctrl/wafl_base_execution_config" \
+    "$USER@${WAFL_DEVICE_IPS[$counter]}:$DEPLOYMENT_LOCATION/$TARGET_NAME" &&
+    ([ "$(ls -A $TARGET_PATH/wafl/dataset/common)" ] && 
+    scp -r -q "$TARGET_PATH/wafl/dataset/common/"* \
+    "$USER@${WAFL_DEVICE_IPS[$counter]}:$DEPLOYMENT_LOCATION/$TARGET_NAME/dataset" || true) &&
```
Not only when the directory is empty, but also when the `scp`  fails, `true` command is executed, making it impossible to detect errors from `scp`. Maybe it would be better to modify the code to use an `if` statement for branching.
This is also the case for other parts (line 53~67).

```suggestion
        (if [ "$(ls -A $TARGET_PATH/wafl/dataset/common)" ]; then 
        scp -r -q "$TARGET_PATH/wafl/dataset/common/"* \
               "$USER@${WAFL_DEVICE_IPS[$counter]}:$DEPLOYMENT_LOCATION/$TARGET_NAME/dataset"
    else
        true
    fi ) &&
```

namitshah replied:
Thank you for detecting the potential bug!
Instead of using multiple if blocks, we can use a variable for detecting when SCP was executed but failed. This will be a compact solution for all the 6 scp commands.

I'll merge the pull request after making this change.

nat0391 reviewed (approved):
I left one comment on the script. This may not be a major issue, but I’d suggest taking it into consideration.
I found no other issues. Thank you for your work.

namitshah committed:
Update deploy.sh
